From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: George Chan <georgechan@google.com>
Date: Thu, 16 Apr 2026 10:13:21 -0700
Subject: [PATCH 01/24] Security bug fix: Prevent USB data changes for
 non-owner users during lockdown mode.

The onStrongAuthRequiredChanged callback now only affects USB data for non-guest users. This prevents USB data from being enabled for guest users when the device enters lockdown mode.

Bug: 466379326
Flag: EXEMPT bugfix
Test: manual, UsbServiceTest
(cherry picked from commit d1d84790170ed14a43e94916b2a72705a28ed120)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:fea130396d83cf4fcf931880fe86acc384ecfb20
Merged-In: Ifa549103a0f454921334aad5687653088609da2d
Change-Id: Ifa549103a0f454921334aad5687653088609da2d

[a17: hand-ported onto e/os a17, which already has sDefaultOperation but still re-enables USB data for any user; UsbServiceTest changes not carried]
(cherry picked from commit 21cc20f26b18d2ea37e7da8ca1f1cc492a5b3558)
---
 .../com/android/server/usb/UsbService.java    | 29 +++++++++++++++----
 1 file changed, 23 insertions(+), 6 deletions(-)

diff --git a/services/usb/java/com/android/server/usb/UsbService.java b/services/usb/java/com/android/server/usb/UsbService.java
index ba7b776420f4..fe7bc5459684 100644
--- a/services/usb/java/com/android/server/usb/UsbService.java
+++ b/services/usb/java/com/android/server/usb/UsbService.java
@@ -89,6 +89,7 @@ import com.android.server.FgThread;
 import com.android.server.LocalServices;
 import com.android.server.SystemServerInitThreadPool;
 import com.android.server.SystemService;
+import com.android.server.pm.UserManagerInternal;
 
 import dalvik.annotation.optimization.NeverCompile;
 
@@ -2000,11 +2001,14 @@ public class UsbService extends IUsbManager.Stub {
         private static final IUsbOperationInternal sDefaultOperation =
                 new IUsbOperationInternal.Default();
         private boolean mLockdownModeStatus;
+        private UserManagerInternal mUserManagerInternal;
 
         StrongAuthTracker(Context context, Looper looper) {
             super(context, looper);
+            mUserManagerInternal = LocalServices.getService(UserManagerInternal.class);
         }
 
+        @SuppressLint("AndroidFrameworkRequiresPermission")
         @Override
         public synchronized void onStrongAuthRequiredChanged(int userId) {
 
@@ -2015,12 +2019,25 @@ public class UsbService extends IUsbManager.Stub {
                 return;
             }
             mLockdownModeStatus = lockDownTriggeredByUser;
-            for (UsbPort port: mPortManager.getPorts()) {
-                enableUsbDataInternal(port.getId(), !lockDownTriggeredByUser,
-                    STRONG_AUTH_OPERATION_ID,
-                    sDefaultOperation,
-                    USB_DISABLE_REASON_LOCKDOWN_MODE,
-                    true);
+            /*
+             * TODO(b/500854486)
+            * Since entering guest mode or switching to non-main user from lockscreen may not
+            * require auth, USB should not be enabled for non-main users. Should remove user switch
+            * access when device is locked down as a follow up.
+            *
+            * For now, we will only re-enable USB for the main user when the device is unlocked.
+            * Secondary users will not have USB enabled even if the device is unlocked, this is
+            * a temporary solution until we remove user switch access from lockscreen under lockdown
+            * mode.
+            */
+            if (lockDownTriggeredByUser || mUserManagerInternal.getMainUserId() == userId) {
+                for (UsbPort port: mPortManager.getPorts()) {
+                    enableUsbDataInternal(port.getId(), !lockDownTriggeredByUser,
+                        STRONG_AUTH_OPERATION_ID,
+                        sDefaultOperation,
+                        USB_DISABLE_REASON_LOCKDOWN_MODE,
+                        true);
+                }
             }
         }
     }
